torch/csrc/jit/codegen/cuda/interface.cpp

```diff
@@ -34,11 +39,129 @@ void fuseGraph(std::shared_ptr<Graph>& graph) {
   getFuserInterface()->fn_fuse_graph(graph);
 }
 
+// [ Note -- type guard logic in CudaFusionGuard ]
+//
+// Type guard confirms that runtime `tensor` complies with profiled
+// `guard_tensor_type`.
+//
+// At this moment, we only do single profiling run, so `guard_tensor_type` has
+// static shape / stride / scalarType. *This might be a little confusing as our
+// implementation is actually more relaxed.
+//
+// Things that we check:
+//   a. identical rank & scalar type
+//   b. stride check:
+//        b.1. identical stride order
+//        b.2. identical contiguity
+//             note that contiguity here is used for tensor collapsing. So extra
+//             attention should be paid to contiguity across size-1 dimensions.
+//   c. size check:
+//        making sure that broadcast semantics are identical. So we want to make
+//        sure a given dimension either are both size-1 for `tensor` &
+//        `guard_tensor_type`, or are both non-size-1.
+//
+bool complyWith(
+    const at::Tensor& tensor,
+    const c10::TensorTypePtr& guard_tensor_type) {
+  // guard broadcast semantics, contiguity & stride order;
+  TORCH_INTERNAL_ASSERT(
+      guard_tensor_type && guard_tensor_type->dim().has_value());
+
+  // check a. if num_dimension check fails or scalar type check fails
+  if (*guard_tensor_type->dim() != static_cast<size_t>(tensor.ndimension()) ||
+      (guard_tensor_type->scalarType().has_value() &&
+       (guard_tensor_type->scalarType().value() != tensor.scalar_type()))) {
+    return false;
+  }
+
+  // TODO: should we get symbolic_size instead and check for size
+  // consistency across tensors as well?
+  const auto& sizes = guard_tensor_type->sizes();
+  const auto& stride_properties = guard_tensor_type->stride_properties();
+
+  const auto& t_sizes = tensor.sizes();
+  const auto& t_strides = tensor.strides();
+  int inner_dim = -1;
+  for (int j = 0; j < *guard_tensor_type->dim(); j++) {
+    // check b. for stride check, we go along dimensions from fastest stride to
+    // slowest stride
+    {
```

Comment:
can this be factored out as helper function instead of an artificial scope here?

Reply:
I can remove the artificial scope. Don't see a good reason to put this out as a helper function if we are not going to use it any where else.